### PR TITLE
Hostless autosaves

### PR DESCRIPTION
**Pull request by o01eg**
Adds autosaves to hostless mode and options to enable/disable it.

Fix issues from http://freeorion.org/forum/viewtopic.php?f=25&t=10830

### CODE REVIEW

File: server/ServerFSM.cpp [hunk 1]
```diff
@@ -2324,6 +2342,9 @@ sc::result ProcessingTurn::react(const ProcessTurn& u) {
         }
     }
 
+    if(server.IsHostless() && GetOptionsDB().Get<bool>("save.server.hostless")) {
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: space between `if` and `(`.

File: server/ServerFSM.cpp [hunk 2]
```diff
@@ -2166,6 +2166,24 @@ sc::result WaitingForTurnEndIdle::react(const SaveGameRequest& msg) {
     return transit<WaitingForSaveData>();
 }
 
+sc::result WaitingForTurnEndIdle::react(const HostlessSave& e) {
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: Is a separate event needed for this? Could you just check in `react(const SaveGameRequest& msg)` if the server is in hostless mode, and if so, don't check for a sender connection or if its player is the host?

Comment (o01eg) [reply server/ServerFSM.cpp]: I've tried but `SaveGameRequest` requires `PlayerConnectionPtr& player_connection` which cann't be `nullptr`.

Comment (o01eg) [reply server/ServerFSM.cpp]: Yes. I suppose if `SaveGameRequest` doesn't have a connection then it was send in hostless mode.

File: util/MultiplayerCommon.cpp [hunk 3]
```diff
@@ -33,6 +33,7 @@ namespace {
         db.Add<std::string>("resource.stringtable.path",    UserStringNop("OPTIONS_DB_STRINGTABLE_FILENAME"),   "");
         db.Add("save.format.binary.enabled",                UserStringNop("OPTIONS_DB_BINARY_SERIALIZATION"),   false);
         db.Add("save.format.xml.zlib.enabled",              UserStringNop("OPTIONS_DB_XML_ZLIB_SERIALIZATION"), true);
+        db.Add("save.server.hostless",                      UserStringNop("OPTIONS_DB_HOSTLESS_SAVE"),          false);
```
Comment (dbenage-cx) [inline util/MultiplayerCommon.cpp]: Perhaps save.auto.hostless.enabled, grouping with other autosave options.

Comment (o01eg) [reply util/MultiplayerCommon.cpp]: Fixed

File: server/ServerFSM.cpp [hunk 4]
```diff
@@ -2324,8 +2324,8 @@ sc::result ProcessingTurn::react(const ProcessTurn& u) {
         }
     }
 
-    if (server.IsHostless() && GetOptionsDB().Get<bool>("save.server.hostless")) {
-        boost::filesystem::path autosave_dir_path = GetServerSaveDir() / "auto";
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: why do you say the auto save subdirectory is unused...?

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: It looks like it checks that the save path is in the save directory, not that it is the save directory.

Comment (o01eg) [reply server/ServerFSM.cpp]: It also checks if `auto` directory exists but don't create it.

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: ok... so create it then? autosaves are supposed to be segregated into the auto directory

Comment (o01eg) [reply server/ServerFSM.cpp]: Ok, I'll look into it.

Comment (o01eg) [reply server/ServerFSM.cpp]: Looks like `boost::filesystem::canonical` in `IsInDir` requires directories to exist. There is a simple solution with `boost::filesystem::weakly_canonical` but it was introduced in boost 1.60

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: > Looks like boost::filesystem::canonical in IsInDir requires directories to exist.

What is the importance / consequence of this?

Comment (o01eg) [reply server/ServerFSM.cpp]: > What is the importance / consequence of this?

It throws error if path not exists: http://www.boost.org/doc/libs/1_58_0/libs/filesystem/doc/reference.html#canonical
Yet I get a little  workaround around it.

File: server/ServerFSM.cpp [hunk 5]
```diff
@@ -2324,6 +2324,20 @@ sc::result ProcessingTurn::react(const ProcessTurn& u) {
         }
     }
 
+    if (server.IsHostless() && GetOptionsDB().Get<bool>("save.auto.hostless.enabled")) {
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp:2327]: I think autosaves are supposed to be done at the end of a turn, not at the start.

File: server/ServerFSM.cpp [hunk 6]
```diff
@@ -2324,6 +2324,20 @@ sc::result ProcessingTurn::react(const ProcessTurn& u) {
...
+        boost::filesystem::path autosave_dir_path = GetServerSaveDir() / "auto";
+        const auto& extension = MP_SAVE_FILE_EXTENSION;
+        // Add timestamp to autosave generated files
+        std::string datetime_str = FilenameTimestamp();
+
+        std::string save_filename = boost::io::str(boost::format("FreeOrion_%04d_%s%s") % server.CurrentTurn() % datetime_str % extension);
+
+        boost::filesystem::path save_path(autosave_dir_path / save_filename);
+
+        PlayerConnectionPtr dummy_connection = nullptr;
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp:2337]: is this temporary local necessary vs. just passing `nullptr` directly?

File: util/MultiplayerCommon.cpp [hunk 7]
```diff
@@ -33,6 +33,7 @@ namespace {
         db.Add<std::string>("resource.stringtable.path",    UserStringNop("OPTIONS_DB_STRINGTABLE_FILENAME"),   "");
         db.Add("save.format.binary.enabled",                UserStringNop("OPTIONS_DB_BINARY_SERIALIZATION"),   false);
         db.Add("save.format.xml.zlib.enabled",              UserStringNop("OPTIONS_DB_XML_ZLIB_SERIALIZATION"), true);
+        db.Add("save.auto.hostless.enabled",                UserStringNop("OPTIONS_DB_HOSTLESS_SAVE"),          false);
```
Comment (geoffthemedio) [inline util/MultiplayerCommon.cpp]: why not true by default?

File: server/ServerFSM.cpp (hunk 6 above)
Comment (o01eg) [reply server/ServerFSM.cpp:2337]: It's necessary as message constructor accepts non-const reference to connection.

File: util/MultiplayerCommon.cpp (hunk 7 above)
Comment (o01eg) [reply util/MultiplayerCommon.cpp]: Made it enabled.

File: server/ServerFSM.cpp (hunk 6 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp:2337]: How long does it expect that reference to be valid? Beyond the lifetime of `dummy_connection`?

Comment (o01eg) [reply server/ServerFSM.cpp:2337]: It's need to be valid until `MessageEventBase` constructor so it don't have to be valid outside its scope.

File: server/ServerFSM.cpp (hunk 5 above)
Comment (o01eg) [reply server/ServerFSM.cpp:2327]: Doesn't turn end in the end of `ProcessingTurn::react(const ProcessTurn&)`?

Comment (geoffthemedio) [reply server/ServerFSM.cpp:2327]: I mean after all players have given orders, it should autosave, then process those orders and the between-turn events and updates. If the save is loaded, players should be able to issue more orders and re-end the turn. At least that's how it should work in single player.

Comment (o01eg) [reply server/ServerFSM.cpp:2327]: I'll try to change it but I wonder which correct state and event sequence to save a game now.

Comment (o01eg) [reply server/ServerFSM.cpp:2327]: I don't see how the game saved after getting all orders and before turn processing. We waits for orders in `TurnOrder` and raise `CheckTurnEndConditions` where the server goes to `ProcessingTurn` state when all orders are received.

Comment (geoffthemedio) [reply server/ServerFSM.cpp:2327]: Singleplayer autosaves are initiated by the player's client when the player clicks end turn, before the actual end turn message is sent, in `HumanClientApp::StartTurn`

File: server/ServerFSM.cpp (hunk 6 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp:2337]: Should probably change `MessageEventBase` to take a `const PlayerConnectionPtr&` then... doesn't seem to need to be a non-const pointer (but does need to be a pointer-to-non-const).

File: server/ServerFSM.cpp (hunk 5 above)
Comment (o01eg) [reply server/ServerFSM.cpp:2327]: The client sends `SAVE_GAME_INITIATE` which translates to `SaveGameRequest` event, but only `WaitingForTurnEndIdle` is a state processes `SaveGameRequest` where the server goes from `WaitingForSaveData` after game saved or from `ProcessingTurn` after `ProcessTurn` has been handled.

I don't see a way how the game can save something between acquiring orders and turn processing.

File: server/ServerFSM.cpp (hunk 6 above)
Comment (o01eg) [reply server/ServerFSM.cpp:2337]: Should it be done in this Pull-Request?

File: server/ServerFSM.cpp (hunk 5 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp:2327]: When the client sends the request, the server is still in `WaitingForTurnEndIdle`.

File: server/ServerFSM.cpp (hunk 6 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp:2337]: I guess it's not necessary.

File: server/ServerFSM.cpp (hunk 5 above)
Comment (o01eg) [reply server/ServerFSM.cpp:2327]: Yes, but it before all players sent their orders.

Comment (o01eg) [reply server/ServerFSM.cpp:2327]: Yes, but other players can send their orders after host.

File: default/stringtables/en.txt [hunk 8]
```diff
@@ -1795,6 +1795,9 @@ Use Binary serialization for saving games. Binary serialization is faster to sav
 OPTIONS_DB_XML_ZLIB_SERIALIZATION
 When saving games with XML serialization, compress most of the XML before writing the file. Compression substantially reduces save file sizes, but may make saves unloadable due to memory requirements to decompress the save data.
 
+OPTIONS_DB_HOSTLESS_SAVE
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: should call this something with `AUTOSAVE` in the name, and put it with the other autosave strings.

File: server/ServerFSM.cpp (hunk 5 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp:2327]: not in single player.

in multiplayer, I think the autosaves are done at the start of a turn. so I guess this timing is OK.

Comment (o01eg) [reply server/ServerFSM.cpp:2327]: Yes, after all orders have been received and processed there will be start of next turn where I push `SaveGameRequest` for hostless autosave.

File: default/stringtables/en.txt (hunk 8 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Renamed and moved.

**Comment by geoffthemedio** [comment]
Does this respect the max autosaves setting?

https://github.com/freeorion/freeorion/blob/master/client/human/HumanClientApp.cpp#L1325

**Comment by o01eg** [comment]
@geoffthemedio No, because it's a client's options. Should it also ported?

**Comment by geoffthemedio** [comment]
Probably should also be an option that can be set on the server. Doesn't need to be in this pull request, though.

These changes work for you as intended?

**Comment by o01eg** [comment]
Yes, they work.

**Pull request merged**